### PR TITLE
add in CFG `ExpressionStatementNode` after every expression statement

### PR DESCRIPTION
**Pull request by d367wang**
To fix https://github.com/typetools/checker-framework/issues/3281, define a new kind of CFG node `ExpressionStatementNode`, for common actions performed after expression statements, e.g. merging conditional stores.

### CODE REVIEW

File: checker/tests/regex/Issue3281.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Test case for Issue 3281:
 // https://github.com/typetools/checker-framework/issues/3281
 
-// @skip-test until the bug is fixed
+
```
Comment (wmdietl) [inline checker/tests/regex/Issue3281.java:4]: I think you're missing an `import java.util.regex.Pattern;`

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 2]
```diff
@@ -2358,7 +2359,12 @@ public Node visitErroneous(ErroneousTree tree, Void p) {
 
     @Override
     public Node visitExpressionStatement(ExpressionStatementTree tree, Void p) {
-        return scan(tree.getExpression(), p);
+        ExpressionTree exprTree = tree.getExpression();
+        scan(exprTree, p);
+        if (!(exprTree instanceof AssignmentTree)) {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:2364]: Why not also for `AssignmentTree`? If needed, can you add a comment that explains the motivation?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MergeOfStoreNode.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+
+/**
+ * A node for the purpose of merging store in dataflow analysis. It usually follows an expression
+ * statement.
+ *
+ * <p>Note: Does not represent any AST structure.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/MergeOfStoreNode.java:17]: Alternatively, we could name it `ExpressionStatementNode` and then it is up to each analysis what they do if it.
At the moment the node name fixes what interpretation we have now, but maybe that could be different for other analyses?

File: checker/tests/regex/Issue3281.java [hunk 4]
```diff
@@ -44,6 +44,7 @@ void m3(String s) {
     void m4(String s, String s2) {
         RegexUtil.isRegex(s);
         if (RegexUtil.isRegex(s2)) {
+            // :: error: (argument.type.incompatible)
             Pattern.compile(s);
         }
     }
```
Comment (wmdietl) [inline checker/tests/regex/Issue3281.java:50]: The PR adds the empty file `dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java` which should be undone.

File: checker/tests/regex/Issue3281.java [hunk 5]
```diff
@@ -44,6 +44,7 @@ void m3(String s) {
     void m4(String s, String s2) {
         RegexUtil.isRegex(s);
         if (RegexUtil.isRegex(s2)) {
+            // :: error: (argument.type.incompatible)
             Pattern.compile(s);
         }
```
Comment (wmdietl) [inline checker/tests/regex/Issue3281.java:49]: Can you please also add these:

````
    void foo(String s1, String s2) {
        bar(RegexUtil.isRegex(s1), 
            // :: error: (argument.type.incompatible)
            Pattern.compile(s1));
        boolean b;
        bar(b = RegexUtil.isRegex(s2),
            // :: error: (argument.type.incompatible)
            Pattern.compile(s2));
    }

    void bar(boolean b, Object o) {}
````

They currently do the right thing, but making sure that the conditional stores don't flow into following arguments would be good (in particular if we change the transfer function for assignment).

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public <R, P> R accept(NodeVisitor<R, P> visitor, P p) {
+        return visitor.visitExpressionStatement(this, p);
+    }
+
+    @Override
+    public String toString() {
+        return "merge store after " + tree.toString();
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java:46]:
```suggestion
        return "expression statement " + tree.toString();
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Objects;
+
+/**
+ * A node for the purpose of merging store in dataflow analysis. It usually follows an expression
+ * statement.
+ *
+ * <p>Note: Does not represent any AST structure.
+ */
+public class ExpressionStatementNode extends Node {
+    /** The expression precedes this MergeOfStoreNode. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java:20]:
```suggestion
    /** The expression constituting this ExpressionStatementNode. */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+
+/**
+ * A node for the purpose of merging store in dataflow analysis. It usually follows an expression
+ * statement.
+ *
+ * <p>Note: Does not represent any AST structure.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java:17]:
```suggestion
 * An expression that is used as a statement.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ */
+public class ExpressionStatementNode extends Node {
+    /** The expression precedes this MergeOfStoreNode. */
+    protected final ExpressionTree tree;
+
+    /** @param t the expression precedes this MergeOfStoreNode */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java:23]:
```suggestion
    /** Construct a ExpressionStatementNode.
     *
     * @param t the expression constituting this ExpressionStatementNode
     */
```

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/ExpressionStatementNode.java:23]: Probably will need reformatting...

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 10]
```diff
@@ -159,6 +159,24 @@
...
+     * @param classDeclarationNode the {@link ClassDeclarationNode} to be visited
+     * @param p the argument of the operation implemented by this visitor
+     * @return the return type of the operation implemented by this visitor, use Void if no return
+     *     type is needed
+     */
     R visitClassDeclaration(ClassDeclarationNode classDeclarationNode, P p);
+
+    //
+    /**
+     * Merges store after an expression statement.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:174]: Please update for new class name

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 11]
```diff
@@ -159,6 +159,24 @@
     // Marker nodes
     R visitMarker(MarkerNode n, P p);
 
-    // Anonymous/inner/nested class declaration within a method
+    /**
+     * Visits anonymous/inner/nested class declaration within a method.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:163]:
```suggestion
     * Visits an anonymous/inner/nested class declaration within a method.
```

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 12]
```diff
@@ -1363,6 +1364,13 @@ protected V getWidenedValue(TypeMirror type, V annotatedValue) {
         return result;
     }
 
+    @Override
+    public TransferResult<V, S> visitExpressionStatement(
+            ExpressionStatementNode n, TransferInput<V, S> vsTransferInput) {
+        S info = vsTransferInput.getRegularStore();
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:1370]:
```suggestion
        // Merge the input
        S info = vsTransferInput.getRegularStore();
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java (hunk 10 above)
Comment (d367wang) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:174]:
```suggestion
     * Actions when visiting an expression used as a statement.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AbstractNodeVisitor.java [hunk 13]
```diff
@@ -380,4 +380,10 @@ public R visitParameterizedType(ParameterizedTypeNode n, P p) {
     public R visitMarker(MarkerNode n, P p) {
         return visitNode(n, p);
     }
+
+    // Merge store after an expression statement
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AbstractNodeVisitor.java:384]:
```suggestion
```

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AbstractNodeVisitor.java:384]: I would remove this comment - it's not what this method does and might not be what an actual subclass does.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 14]
```diff
@@ -159,6 +159,25 @@
...
+     * Visits an anonymous/inner/nested class declaration within a method.
+     *
+     * @param classDeclarationNode the {@link ClassDeclarationNode} to be visited
+     * @param p the argument of the operation implemented by this visitor
+     * @return the return type of the operation implemented by this visitor, use Void if no return
+     *     type is needed
+     */
     R visitClassDeclaration(ClassDeclarationNode classDeclarationNode, P p);
+
+    //
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:172]:
```suggestion
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 15]
```diff
@@ -159,6 +159,25 @@
...
     R visitMarker(MarkerNode n, P p);
 
-    // Anonymous/inner/nested class declaration within a method
+    /**
+     * Visits an anonymous/inner/nested class declaration within a method.
+     *
+     * @param classDeclarationNode the {@link ClassDeclarationNode} to be visited
+     * @param p the argument of the operation implemented by this visitor
+     * @return the return type of the operation implemented by this visitor, use Void if no return
+     *     type is needed
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:168]:
```suggestion
     * @return the return value of the operation implemented by this visitor
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 16]
```diff
@@ -159,6 +159,25 @@
     // Marker nodes
     R visitMarker(MarkerNode n, P p);
 
-    // Anonymous/inner/nested class declaration within a method
+    /**
+     * Visits an anonymous/inner/nested class declaration within a method.
+     *
+     * @param classDeclarationNode the {@link ClassDeclarationNode} to be visited
+     * @param p the argument of the operation implemented by this visitor
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:166]:
```suggestion
     * @param p the argument for the operation implemented by this visitor
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java (hunk 15 above)
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:168]: The documentation is about the value that is returned, not the type.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 17]
```diff
@@ -159,6 +159,25 @@
...
+     * @return the return type of the operation implemented by this visitor, use Void if no return
+     *     type is needed
+     */
     R visitClassDeclaration(ClassDeclarationNode classDeclarationNode, P p);
+
+    //
+    /**
+     * Visits a non-syntactical {@link ExpressionStatementNode}, which is appended to the
+     * syntactical nodes of an expression statement.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:175]:
```suggestion
     * Visits an expression that is used as a statement. This node is a marker after the expression node(s).
```

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:175]: Not sure how to word this. Should something about the order in the CFG should also go into the class-doc for ExpressionStatementNode?
Please also add a comment about the new node and its location to the manual: https://github.com/eisop/checker-framework/blob/master/dataflow/manual/content.tex

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 18]
```diff
@@ -159,6 +159,25 @@
...
+     *
+     * @param n the {@link ExpressionStatementNode} to be visited
+     * @param p the argument of the operation implemented by this visitor
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:178]:
```suggestion
     * @param p the argument for the operation implemented by this visitor
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java [hunk 19]
```diff
@@ -159,6 +159,25 @@
...
+     * @return the return type of the operation implemented by this visitor, use Void if no return
+     *     type is needed
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/NodeVisitor.java:180]:
```suggestion
     * @return the return value of the operation implemented by this visitor
```

**Review by wmdietl — APPROVED** [review]
Thanks for polishing this PR! I have a few small comments and suggestions, but all this looks good. We can ignore the javadoc CI failure.
Please close the PR here and move it to eisop, where we can merge it separately from https://github.com/eisop/checker-framework/pull/163
File an issue if you want me to make the manual changes separately.

**Comment by d367wang** [comment]
Moved it to `eisop`

**Pull request closed without merging**